ognl/src/main/java/ognl/OgnlRuntime.java

```diff
@@ -668,27 +668,32 @@ public static Object invokeMethod(Object target, Method method, Object[] argsArr
         }
 
         // only synchronize method invocation if it actually requires it
-
-        synchronized (method) {
-            methodAccessCacheValue = _methodAccessCache.get(method);
-            if (methodAccessCacheValue == null) {
-                if (!Modifier.isPublic(method.getModifiers()) || !Modifier.isPublic(method.getDeclaringClass().getModifiers())) {
-                    var obj = Modifier.isStatic(method.getModifiers()) ? null : target;
-                    if (method.canAccess(obj)) {
-                        methodAccessCacheValue = Boolean.FALSE;
-                        _methodAccessCache.put(method, methodAccessCacheValue);
+        methodAccessCacheValue = _methodAccessCache.get(method);
+        // double null check to avoid synchronizing on the method
+        if (methodAccessCacheValue == null) {
+            synchronized (method) {
+                methodAccessCacheValue = _methodAccessCache.get(method);
+                if (methodAccessCacheValue == null) {
+                    if (!Modifier.isPublic(method.getModifiers()) || !Modifier.isPublic(method.getDeclaringClass().getModifiers())) {
+                        var obj = Modifier.isStatic(method.getModifiers()) ? null : target;
+                        if (method.canAccess(obj)) {
+                            methodAccessCacheValue = Boolean.FALSE;
+                            _methodAccessCache.put(method, methodAccessCacheValue);
+                        } else {
+                            methodAccessCacheValue = Boolean.TRUE;
+                            _methodAccessCache.put(method, methodAccessCacheValue);
+                        }
                     } else {
-                        methodAccessCacheValue = Boolean.TRUE;
+                        methodAccessCacheValue = Boolean.FALSE;
                         _methodAccessCache.put(method, methodAccessCacheValue);
                     }
-                } else {
-                    methodAccessCacheValue = Boolean.FALSE;
-                    _methodAccessCache.put(method, methodAccessCacheValue);
                 }
+                syncInvoke = Boolean.TRUE.equals(methodAccessCacheValue);
```

Comment:
This and the next assignment in the `else` block can be extracted out of the `if/else` block, wdyt?